### ISSUE TITLE
Allow users to specify the LLM endpoint (APIs) for other models

### ISSUE THREAD

**Issue by nmcbride** [warpdotdev/Warp#4687]
### Describe the solution you'd like?

Allow us to specify the LLM endpoint to use for local models or other online ones.

Would like to point warp to either a local LLM or alternative onlien LLM. 

### How important is this feature to you?

3 (Fairly important)

### Warp Internal (ignore) - linear-label:39cc6478-1249-4ee7-950b-c428edfeecd1

None

**Comment by dannyneira** [comment]
Hi @nmcbride Thanks for submitting.

We already have a request tracking this over at https://github.com/warpdotdev/Warp/issues/3779

If you could please give it a 👍 and subscribe to that request, it would help us gauge interest. We'll post any updates there, so stay tuned! 

We'll be closing this as a duplicate.

**Comment by sintaxx** [comment]
I already have subscriptions to Claude and ChatGPT and would love to be able to plug them in to warp.

**Comment by dwSun** [comment]
As a commercial company, paid AI usage is Warp's profit model. So Warp may not have much motivation to do this on its own. However, considering that deploying an LLM model is already a highly technical behavior. Having this function can be another gimmick for promoting products. In comparison, Warp's market share may further increase. Therefore, please consider providing this function and use it to promote products.

**Comment by drakzonbm** [comment]
I'd love to see this feature and I don't think it would interfere that much with the commercial nature of WARP local LLM's will always be a little behind when it comes to intelligence plenty of people will want to use a commercial LLM at times even if this is an option I think

**Comment by thezacharytaylor** [comment]
I'd love to plug Mistral's new Devstral into Warp and try it out so this would be great if we could do this and specify the model as well via a text field.

**Comment by nmcbride** [comment]
> Warp can make money by offering a 5$/mo subscription or similar for self hosted models. being a token broker middleman as business model doesn't pay.

Yes they have many ways they could go with it and still make money but so far they have shown themselves unwilling to do so unless you pay for enterprise.

I've been using wave terminal instead. It's a bit different but they have AI baked in and support for local models out of the box. Not as polished thought but they are still in Beta.

**Comment by juanmiguelra2005** [comment]
Warp without local LLM is useless. Data protection is of primary importance. I delete the app from my computer.

**Comment by Anki-code1** [comment]
> ### Discord username (optional)
> 
> 
> 
> _No response_
> 
> 
> 
> ### Describe the solution you'd like?
> 
> 
> 
> Allow us to specify the LLM endpoint to use for local models or other online ones.
> 
> 
> 
> Would like to point warp to either a local LLM or alternative onlien LLM. 
> 
> 
> 
> ### Is your feature request related to a problem? Please describe.
> 
> 
> 
> _No response_
> 
> 
> 
> ### Additional context
> 
> 
> 
> _No response_
> 
> 
> 
> ### How important is this feature to you?
> 
> 
> 
> 3 (Fairly important)
> 
> 
> 
> ### Warp Internal (ignore) - linear-label:39cc6478-1249-4ee7-950b-c428edfeecd1
> 
> 
> 
> None

#1724

**Comment by melkordoran** [comment]
https://www.warp.dev/enterprise
Can we use our own LLM?
Warp supports LLMs hosted by your company, please check with your Warp contact to determine if your hosted model is compliant with Agent Mode API (Anthropic and OpenAI are currently supported). Note that at the moment, AI requests are still proxied through Warp Server. Zero Data Retention Guarantees are automatically enabled when providing your own LLM endpoint.

Given your stance on this will you be changing product development direction to allow for end users to specify their own locally hosted LLM?

Knowing that my BYOK is stored within app and used by your proxy server still raises alarm bells to me which I have no control over at a typical users price point unless you restructure to allow for local LLMs which similar product offerings allow this make Warp no longer viable.

I will not be renewing my subscription until this changes.

You can tell Zach and team to stop e-mailing me, this is my response to your outreach.

**Comment by dannyneira** [comment]
Commenting with details from duplicate #9241

## Summary
Add support for OpenAI-compatible BYOK endpoints by letting users configure a base URL and custom model ID, so Warp agents can use OpenRouter without requiring a separate first-class provider integration.

### Problem
Warp already supports a curated set of built-in models and limited BYOK flows, but there is still no straightforward way for a non-enterprise user to point Warp at an OpenAI-compatible endpoint such as OpenRouter.

That leaves a large gap for users who:

already pay for OpenRouter or another OpenAI-compatible gateway,
want to choose from more models than Warp exposes directly,
need a single routed endpoint for availability, cost control, or provider flexibility,
or cannot justify paying for both Warp-managed tokens and a separate provider subscription.
This request is intentionally narrower than a full "bring any model/provider" system. The minimal feature needed here is:

a configurable OpenAI-compatible base URL,
an API key for that endpoint,
a custom model identifier string,
and using that configuration for Warp agent requests when selected.
If Warp supported just that, OpenRouter would work immediately, and the same surface would also unlock many compatible gateways (LiteLLM, internal proxies, Azure-compatible wrappers, etc.) without requiring Warp to add bespoke integrations for each one.

#### Reproduction steps or desired workflow
Open Settings > AI.
Add a new provider type such as OpenAI-compatible.
Configure:
Base URL, for example https://openrouter.ai/api/v1
API key
Model ID, for example anthropic/claude-opus-4.1 or another OpenRouter model slug
Select that provider/model in an Agent profile or model picker.
Run Warp agent features normally in the terminal.

**Comment by cedric-albeke** [comment]
+1 for generic OpenAI-compatible endpoints. A simple provider config with name/base_url/api_key/model would unlock Sub, LiteLLM, OpenRouter-compatible routers, Azure, local gateways, Kimi/GLM, etc. without per-provider UI work.

**Comment by zerx-lab** [comment]
For folks tracking this — there's a community fork **[OpenWarp](https://github.com/zerx-lab/warp)** that already implements BYOP on top of upstream Warp:

- Plug in **any** OpenAI / Anthropic / Gemini / Ollama / DeepSeek–compatible endpoint via Base URL + API key (5 native protocols built on [genai](https://github.com/jeremychone/rust-genai) 0.6)
- Credentials stored **locally only**, never routed through Warp's cloud
- System prompts are minijinja templates, fully editable (8 model-family templates ported from opencode)
- Blocks / Workflows / Keymaps / AI command suggestions all preserved (continuously merged with upstream)
- models.dev integration for quickly adding model entries

Not affiliated with Warp Inc. — just sharing in case it helps while the official feature is being considered.
Source: https://github.com/zerx-lab/warp

**Comment by bholmesdev** [comment]
Thanks, this maps closely to the direction we’re taking. @zachlloyd  shared that we plan to support arbitrary OpenAI-compatible endpoints for BYOK in the next 1–2 weeks, while also working toward fully client-side local model support and ACP client support. Details here: https://github.com/warpdotdev/warp/discussions/9619#discussioncomment-16831279

**Comment by petradonka** [comment]
Thanks for raising this. Warp now supports custom inference endpoints compatible with the OpenAI Chat Completions API.

This lets you connect Warp to providers, routers, gateways, internal endpoints, or self-hosted inference setups that expose an OpenAI-compatible API.

Docs: https://docs.warp.dev/agent-platform/inference/custom-inference-endpoint/

Closing this as completed now that the core custom endpoint request is covered. If a specific model or provider doesn’t work through the custom endpoint flow, please open a new focused issue with the provider, endpoint format, model name, and exact error.
